### ISSUE TITLE
Black screen on launch

### ISSUE THREAD

**Issue by torvim** [JLErvin/berry#10]
When I launch berry (no dm), it launches into a black screen, without a cursor.

**Comment by nick92** [comment]
also getting a similar issue on load, but cursor is showing for me, installed via yay berry-git and using lightdm

**Comment by Djamy** [comment]
Hello,

Same black screen after testing it on a VM with Ubuntu 18.04.
Do you have tips to help ? Logs somewhere ?
Thanks.

**Comment by ekomizer** [comment]
Do a `chmod +x` on the `autostart` file.

**Comment by Djamy** [comment]
Unfortunately that didn't help

**Comment by JLErvin** [comment]
Try redirecting the output to a log file and post the output here. My guess is that berry isn't starting correctly but this should help narrow it down. Start berry with the command `berry > /tmp/log.txt 2>&1` and upload the output here and I'll take a look.

**Comment by Djamy** [comment]
Hi, thanks for your answer.

Here is the output of the berry command : 
`berry: font not specified, loading default font
berry: Successfully opened display
Fatal request. request code=2, error code=10
X Error of failed request:  BadAccess (attempt to access private resource denied)
  Major opcode of failed request:  2 (X_ChangeWindowAttributes)
  Serial number of failed request:  22
  Current serial number in output stream:  24`

I don't know why it attempt to access private resource though.
I got this error by running berry directly in my terminal, so X server was already started and used.

**Comment by JLErvin** [comment]
@Djamy `berry` cannot be started inside of another xsession. Does this error persist if you start `berry` by running `startx` inside of `tty` or by starting from a display manager.

**Comment by jolupa** [comment]
Hi!
Same problem here, after installing berry and launch through a multiple .xinitrc, tryied only berry in the .xinitrc, but it didn't work black screen, no color background (I put a color background on autostart) and didn't work.
But if I launch Google Chrome is launched. But if I launch the terminal (xfce4-terminal) berry crashes and I go back to tty.
The autostart file is executable.
Any help?

**Comment by LucaT1** [comment]
Okay, so I believe that this is a problem with fonts. In fact, I tried to log the output as described here:

> Try redirecting the output to a log file and post the output here. My guess is that berry isn't starting correctly but this should help narrow it down. Start berry with the command `berry > /tmp/log.txt 2>&1` and upload the output here and I'll take a look.

and it threw a segfault when drawing text. I think this is because on voidlinux(what I'm currently using) no font is available when installing the xorg package. So I tried installing `font-firacode` and started the X session and it worked. Hope this helps somebody.

**Comment by jolupa** [comment]
> Okay, so I believe that this is a problem with fonts. In fact, I tried to log the output as described here:
> 
> > Try redirecting the output to a log file and post the output here. My guess is that berry isn't starting correctly but this should help narrow it down. Start berry with the command `berry > /tmp/log.txt 2>&1` and upload the output here and I'll take a look.
> 
> and it threw a segfault when drawing text (Same behavior as #229. I think this is because on voidlinux(what I'm currently using) no font is available when installing the xorg package. So I tried installing `font-firacode` and started the X session and it worked. hope it helps.

Um! I have xfonts installed and the blank screen remains... I can launch program but nothing about the background color, all black....

**Comment by LucaT1** [comment]
Aw yeah tough the topic was different. Have you tried hsetroot or feh to set the background

**Comment by jolupa** [comment]
Yeah is set with hsetroot... But nothing! :cry:

**Comment by JLErvin** [comment]
@jolupa can you post your `.xinitrc`, your `autostart`, and a log file from crashing

**Comment by jolupa** [comment]
Hi! sorry for the delay.
Here are my .xinitrc and autostart
Thanks.
.xinitrc
```
#!/bin/sh

#Xfce Default
session=${1:-xfce}

case $session in
	berry			)
		exec berry;;
	xfce			)
		exec ck-launch-session startxfce4;;
	*					)
		exec $1;;
esac
```

Autostart
```
#!/bin/bash

source $HOME/.bashrc
sxhkd -c $HOME/.config/berry/sxhkdrc &
$HOME/.fehbg

setxkbmap -layout us -variant altgr-intl

# start bars
#tint2 -c $HOME/dots/tint2/time &
#tint2 -c $HOME/dots/tint2/desktop &

berryc top_gap 0

berryc border_width 0
berryc inner_border_width 5
berryc title_height 20

berryc focus_color 292d3e
berryc unfocus_color 292d3e
berryc inner_focus_color b89ca8
berryc inner_unfocus_color 343d41
berryc text_focus_color 2c3539
berryc text_unfocus_color 565679
```

**Comment by JLErvin** [comment]
@jolupa can you post a log file from the crash? Simply start `berry` with:

`berry > /tmp/log.txt 2>&1`

**Comment by jolupa** [comment]
@JLErvin Here's the log file, not any error except for the font.
I think that I made something wrong. Now I have a background, the one I want, but the mouse pointer is not showing 'till the first window is open.
Don't know... :stuck_out_tongue: 
[log.txt](https://github.com/JLErvin/berry/files/3400918/log.txt)

**Comment by JLErvin** [comment]
@jolupa this is the whole log file?! That isn't even an error - that's just `berry` letting us know that you didn't specify a font via the command line and it will instead load the default font (i.e. Monospace 10) from `config.h`. If there's more to the log please post it.

**Comment by jolupa** [comment]
@JLErvin no that's all. Viewing the log I think that the problem, maybe, is in my side. As I told you beofore, now I have background and the only weird thing is that I don't have cursor till the first window is open.

**Comment by JLErvin** [comment]
@jolupa oh, this is the correct behavior.

**Comment by Djamy** [comment]
@JLErvin Hello, I'm still struggling with this and I cannot start berry from gdm3..
I traced the syslog and found a segfault from libc-2.27.so (and a fantastic "fuck me" in it :rofl: )

Could you help to understand what is wrong and why I can't use your wm, please ?

[berry_syslog.log](https://github.com/JLErvin/berry/files/3610983/berry_syslog.log)

**Comment by JLErvin** [comment]
@Djamy I recently added a fix that was causing crashes on the [development branch](https://github.com/JLErvin/berry/tree/blackberry). Can you try cloning and remaking the project from here and see if it fixes the issue.

P.S. I need to screen my commits better next time haha

**Comment by Djamy** [comment]
@JLErvin Thank you, it seems to solve my problem.
Now I'm facing a kind of blank screen but it is supposed to be the correct behavior as you said, still how can I have the mouse cursor or even launch programs ?

**Comment by JLErvin** [comment]
@Djamy enough people were requesting a default cursor that I just added one as of [0.1.2](https://github.com/JLErvin/berry/tree/0.1.2)

With regards to launching applications this is something you will have to do on your end. Following the traditional unix philosophy, `berry` doesn't handle something like launching applications for you. If you are looking for good application launchers I personally use [dmenu](https://tools.suckless.org/dmenu/) although I know many people that also use [rofi](https://github.com/davatorium/rofi). 

To set a background I recommend a program like [hsetroot](https://github.com/himdel/hsetroot)

**Comment by vaskark** [comment]
I'm experiencing the same problem on Ubuntu 19.04. Starting from gdm just loops me back there. Starting from CLI I receive:

`berry: font not specified, loading default font`

That's it. It seems like this has been solved for others above, but I'm still having trouble.

**Comment by Djamy** [comment]
@JLErvin Okay thank you, I'll see how I can handle this on myself.
@vaskark You should try to checkout the blackberry branch : 

> 
> 
> @Djamy I recently added a fix that was causing crashes on the [development branch](https://github.com/JLErvin/berry/tree/blackberry). Can you try cloning and remaking the project from here and see if it fixes the issue.
> 
> P.S. I need to screen my commits better next time haha

**Comment by vaskark** [comment]
@Djamy Ok. I'm on dev branch. Loads fine now, except for font.

**Comment by JLErvin** [comment]
@vaskark what happens with the font? If you are getting that message but the fonts are still drawn then there is nothing wrong - it must means you aren't specifying a font at startup with the `-f` flag.

**Comment by vaskark** [comment]
@JLErvin My apologies. I didn't have it formated properly in my autostart above. All is fine now :)

It's a neat wm. Congrats, and thank you.

P.S. I'm starting berry from gdm. Is there an exit command I can set in sxhkdrc?

**Comment by JLErvin** [comment]
@vaskark This has also been a requested feature so I'll add a dedicated command in a future update. Until then that works fine, however.

**Comment by JLErvin** [comment]
@vaskark `berryc quit` has been implemented as of 0.1.3. Closing this issue due to inactivity.
